File: src/incidents/create.js
/**
 * Lambda function: Crear nuevo incidente
 * Endpoint: POST /incidents
 * Requiere: Autenticación (todos los roles pueden crear incidentes)
 */
import { v4 as uuidv4 } from 'uuid';
import { putItem, getUserById } from '../utils/dynamodb.js';
import { requireAuth } from '../utils/auth.js';
import { created, badRequest, unauthorized, internalError } from '../utils/responses.js';
import { validateCreateIncident, sanitizeString } from '../utils/validators.js';
import { uploadMultipleImages } from '../utils/s3.js';

const INCIDENTS_TABLE = process.env.INCIDENTS_TABLE;

/**
 * Crea un nuevo incidente en el sistema
 * 
 * Request Body:
 * {
 *   "title": "Luz fundida en Aula 302",
 *   "description": "La luz fluorescente está fundida...",
 *   "category": "electricidad",
 *   "priority": "medium",
 *   "location": {
 *     "building": "Pabellón A",
 *     "floor": 3,
 *     "room": "302",
 *     "specificLocation": "Aula de clases"
 *   },
 *   "images": ["data:image/jpeg;base64,...", ...]
 * }
 */
export const handler = async (event) => {
  try {
    // Verificar autenticación
    let user;
    try {
      user = requireAuth(event);
    } catch (error) {
      return unauthorized(error.message);
    }

    // Parsear body
    const body = JSON.parse(event.body || '{}');

    // Validar datos
    const validation = validateCreateIncident(body);
    if (!validation.valid) {
      return badRequest('Datos del incidente inválidos', {
        errors: validation.errors
      });
    }

    const { title, description, category, priority, location, images } = body;

    // Generar ID único
    const incidentId = `inc_${uuidv4().replace(/-/g, '').substring(0, 12)}`;
    const currentTimestamp = Date.now();

    // Construir objeto de incidente
    const incidentItem = {
      incidentId,
      title: sanitizeString(title),
      description: sanitizeString(description),
      category,
      priority,
      status: 'pending',
      location: {
        building: sanitizeString(location.building),
        floor: location.floor,
        room: sanitizeString(location.room),
        specificLocation: location.specificLocation 
          ? sanitizeString(location.specificLocation) 
          : null
      },
      images: [],
      reportedBy: user.userId,
      assignedTo: null,
      createdAt: currentTimestamp,
      updatedAt: currentTimestamp,
      resolvedAt: null,
      comments: []
    };

    // Subir imágenes a S3 si existen
    let imageUrls = [];
    if (images && images.length > 0) {
      try {
        const uploadedImages = await uploadMultipleImages(images, incidentId);
        
        // Guardar las keys de S3 en el incidente
        incidentItem.images = uploadedImages.map(img => img.s3Key);
        
        // Guardar las URLs públicas para la respuesta
        imageUrls = uploadedImages.map(img => img.publicUrl);
        
      } catch (error) {
        console.error('Error subiendo imágenes:', error.message);
        // Continuar sin imágenes en caso de error
        incidentItem.images = [];
      }
    }

    // Guardar incidente en DynamoDB
    await putItem(INCIDENTS_TABLE, incidentItem);

    // Obtener datos del usuario que reporta
    const reporter = await getUserById(user.userId);
    const reporterInfo = {
      userId: user.userId,
      name: reporter ? reporter.name : 'Usuario',
      email: reporter ? reporter.email : ''
    };

    // TODO: Enviar notificación WebSocket a administradores
    // await broadcastNewIncident(incidentItem, reporterInfo);

    // TODO: Enviar notificación SNS
    // await sendSNSNotification(incidentItem);

    // Preparar respuesta
    const responseData = {
      incidentId,
      title: incidentItem.title,
      description: incidentItem.description,
      category: incidentItem.category,
      priority: incidentItem.priority,
      status: incidentItem.status,
      location: incidentItem.location,
      reportedBy: reporterInfo,
      createdAt: currentTimestamp,
      imageUrls
    };

    return created('Incidente creado exitosamente', responseData);

  } catch (error) {
    if (error instanceof SyntaxError) {
      return badRequest('Formato JSON inválido');
    }

    console.error('Error en create incident:', error);
    return internalError('Error al crear incidente');
  }
};

File: src/incidents/create.py
"""
Lambda function: Crear nuevo incidente
Endpoint: POST /incidents
Requiere: Autenticación (todos los roles pueden crear incidentes)
"""
import json
import os
import time
import uuid
from typing import Dict, Any

# Importar utilidades
import sys
sys.path.append(os.path.join(os.path.dirname(__file__), '..'))

from websocket.broadcast import *
from utils.responses import created, bad_request, unauthorized, internal_error
from utils.dynamodb import put_item, get_user_by_id, INCIDENTS_TABLE
from utils.validators import CreateIncidentSchema, sanitize_string
from utils.auth import require_auth
from utils.s3_helper import upload_multiple_images
from pydantic import ValidationError


def handler(event: Dict[str, Any], context: Any) -> Dict[str, Any]:
    """
    Crea un nuevo incidente en el sistema
    
    Request Body:
    {
        "title": "Luz fundida en Aula 302",
        "description": "La luz fluorescente está fundida...",
        "category": "electricidad",
        "priority": "medium",
        "location": {
            "building": "Pabellón A",
            "floor": 3,
            "room": "302",
            "specificLocation": "Aula de clases"
        },
        "images": ["data:image/jpeg;base64,...", ...]
    }
    
    Response:
    {
        "success": true,
        "message": "Incidente creado exitosamente",
        "data": {
            "incidentId": "inc_xyz789",
            "title": "...",
            "status": "pending",
            "createdAt": 1700000000000,
            "imageUrls": ["https://..."]
        }
    }
    """
    try:
        # Verificar autenticación
        try:
            user = require_auth(event)
        except ValueError as e:
            return unauthorized(str(e))
        
        # Parsear body
        body = json.loads(event.get('body', '{}'))
        
        # Validar datos con Pydantic
        try:
            incident_data = CreateIncidentSchema(**body)
        except ValidationError as e:
            errors = {}
            for error in e.errors():
                field = '.'.join(str(loc) for loc in error['loc'])
                errors[field] = error['msg']
            return bad_request(
                message="Datos del incidente inválidos",
                details=errors
            )
        
        # Generar ID único
        incident_id = f"inc_{uuid.uuid4().hex[:12]}"
        current_timestamp = int(time.time() * 1000)
        
        # Construir objeto de incidente
        incident_item = {
            'incidentId': incident_id,
            'title': sanitize_string(incident_data.title),
            'description': sanitize_string(incident_data.description),
            'category': incident_data.category,
            'priority': incident_data.priority,
            'status': 'pending',
            'location': {
                'building': sanitize_string(incident_data.location.building),
                'floor': incident_data.location.floor,
                'room': sanitize_string(incident_data.location.room),
                'specificLocation': sanitize_string(incident_data.location.specificLocation) 
                    if incident_data.location.specificLocation else None
            },
            'images': [],
            'reportedBy': user['userId'],
            'assignedTo': None,
            'createdAt': current_timestamp,
            'updatedAt': current_timestamp,
            'resolvedAt': None,
            'comments': []
        }
        
        # Subir imágenes a S3 si existen
        image_urls = []
        if incident_data.images and len(incident_data.images) > 0:
            try:
                uploaded_images = upload_multiple_images(
                    incident_data.images,
                    incident_id
                )
                
                # Guardar las keys de S3 en el incidente
                incident_item['images'] = [s3_key for s3_key, _ in uploaded_images]
                
                # Guardar las URLs públicas para la respuesta
                image_urls = [url for _, url in uploaded_images]
                
            except Exception as e:
                print(f"Error subiendo imágenes: {str(e)}")
                # Continuar sin imágenes en caso de error
                incident_item['images'] = []
        
        # Guardar incidente en DynamoDB
        put_item(INCIDENTS_TABLE, incident_item)
        
        # Obtener datos del usuario que reporta
        reporter = get_user_by_id(user['userId'])
        reporter_info = {
            'userId': user['userId'],
            'name': reporter.get('name', 'Usuario') if reporter else 'Usuario',
            'email': reporter.get('email', '') if reporter else ''
        }
        
        # TODO: Enviar notificación WebSocket a administradores
        broadcast_new_incident(incident_item, reporter_info)
        
        # TODO: Enviar notificación SNS
        # send_sns_notification(incident_item)
        
        # Preparar respuesta
        response_data = {
            'incidentId': incident_id,
            'title': incident_item['title'],
            'description': incident_item['description'],
            'category': incident_item['category'],
            'priority': incident_item['priority'],
            'status': incident_item['status'],
            'location': incident_item['location'],
            'reportedBy': reporter_info,
            'createdAt': current_timestamp,
            'imageUrls': image_urls
        }
        
        return created(
            message="Incidente creado exitosamente",
            data=response_data
        )
    
    except json.JSONDecodeError:
        return bad_request("Formato JSON inválido")
    
    except Exception as e:
        print(f"Error en create incident: {str(e)}")
        import traceback
        traceback.print_exc()
        return internal_error(f"Error al crear incidente: {str(e)}")